client/client.go

```diff
@@ -486,3 +546,18 @@ func hash(pwd, salt string, iterations int) string {
 	}
 	return fmt.Sprintf("%x", hash)
 }
+
+func getTCPConn(pool util.Pool) (*net.TCPConn, error) {
+	poolConn, err := pool.Get()
+	if err != nil {
+		return nil, err
+	}
+	defer poolConn.Close()
```

Comment:
Aren't you checking in the conn as soon as this function returns, thereby invalidating the connection?  Typically you will use conn within a closure:

getTCPConn(func(conn) {
  conn.Write...
  conn.Read...
})

So the connection is implicitly checked out and checked in for the closure function.

Reply:
That line shouldn't be there at all, my bad (I mean, the pool should not be closed there. In fact, every place that calls this function should `defer tcpConn.Close()`)

But I don't understand what you mean with the closure. Having a valid pool, requesting a connection, doing all operations and then closing that particular connection should be valid, isn't it?

Reply:
You never close a connection, you check it out, use it to write a request, read a reply and then check it back in.  The pool's Close() call is named poorly; it's actually checking the connection back into the pool, not closing it.